server/src-lib/Hasura/RQL/DML/RemoteJoin.hs

```diff
@@ -0,0 +1,446 @@
+                           (extractAtPath rest)
+                           (AO.lookup (G.unName h) o)
+          AO.Array arr -> AO.array <$> mapM (extractAtPath path) (toList arr)
+          _            -> throw500 $ "expecting array or object in remote response at path " <> T.pack (show path)
+
+-- | Fold nested 'FieldCall's into a bare 'Field', inserting the passed
+-- selection set at the leaf of the tree we construct.
+fieldCallsToField
+  :: MonadError QErr m
+  => [G.Argument]
+  -> Map.HashMap G.Variable G.Value
+  -> [G.Field]
+  -- ^ Inserted at leaf of nested FieldCalls
+  -> G.Alias
+  -- ^ Top-level name to set for this Field
+  -> NonEmpty FieldCall
+  -> m G.Field
+fieldCallsToField rrArguments variables finalSelSet topAlias =
+  fmap (\f -> f{G._fAlias = Just topAlias}) . nest
+  where
+    -- almost: `foldr nest finalSelSet`
+    nest ((FieldCall name remoteArgs) :| rest) = do
+      templatedArguments <- createArguments variables remoteArgs
+      (args, selSet) <- case NE.nonEmpty rest of
+            Just f -> do
+              s <- nest f
+              pure (templatedArguments, pure s)
+            Nothing ->
+              let argsToMap = Map.fromList . map (G._aName &&& G._aValue)
+                  arguments = map (uncurry G.Argument) $ Map.toList $
+                              Map.unionWith mergeValue
+                                (argsToMap rrArguments)
+                                (argsToMap templatedArguments)
+              in pure (arguments, finalSelSet)
+      pure $ G.Field Nothing name args [] $ map G.SelectionField selSet
+
+-- This is a kind of "deep merge".
+-- For e.g. suppose the input argument of the remote field is something like:
+-- `where: { id : 1}`
+-- And during execution, client also gives the input arg: `where: {name: "tiru"}`
+-- We need to merge the input argument to where: {id : 1, name: "tiru"}
+mergeValue :: G.Value -> G.Value -> G.Value
+mergeValue lVal rVal = case (lVal, rVal) of
+  (G.VList (G.ListValueG l), G.VList (G.ListValueG r)) ->
+    G.VList $ G.ListValueG $ l <> r
+  (G.VObject (G.ObjectValueG l), G.VObject (G.ObjectValueG r)) ->
+    let fieldsToMap = Map.fromList . map (G._ofName &&& G._ofValue)
+    in G.VObject $ G.ObjectValueG $ map (uncurry G.ObjectFieldG) $ Map.toList $
+       Map.unionWith mergeValue (fieldsToMap l) (fieldsToMap r)
+  (l, _) -> l -- FIXME:- throw error for merging non-lists and non-objects
+
+fieldsToRequest :: G.OperationType -> [G.Field] -> GQLReqParsed
+fieldsToRequest opType gfields =
+  GQLReq
+     { _grOperationName = Nothing
+     , _grQuery =
+         GQLExecDoc
+           [ G.ExecutableDefinitionOperation
+               (G.OperationDefinitionTyped
+                 ( emptyOperationDefinition
+                     {G._todSelectionSet = map G.SelectionField gfields}
+                 )
+               )
+           ]
+     , _grVariables = Nothing -- TODO: Put variables in here?
```

Comment:
What do you think about this `TODO`? I would need to look closer to have any insight

Reply:
I know what this is: if you have variables used in your remote joins field (from the client), then after validation phase it is expanded i.e no longer a variable.

But this is bad from an operation perspective because the remote server will see the expanded execution (so the variable values are out in the open). I tried to fix this in heterogenous requests PR here: https://github.com/hasura/graphql-engine/pull/3153/files#diff-ae73919a18ce8e018f747703c760d5c5R456

Basically, I un-expand the variables and plug it back.